discovery/file/file_test.go

```diff
@@ -15,158 +15,456 @@ package file
+		tgs = append(tgs, t.tgs[k])
 	}
+	return tgs
+}
 
-	// Test file is ready so stop draining the discovery channel.
-	// It contains two target groups.
-	close(fileReady)
-	<-drainReady
-	newf.WriteString(" ") // One last meaningless write to trigger fsnotify and a new loop of the discovery service.
+func (t *testRunner) requireUpdate(ref time.Time, expected []*targetgroup.Group) {
+	t.Helper()
 
-	timeout := time.After(15 * time.Second)
-retry:
 	for {
 		select {
-		case <-timeout:
-			if expect {
-				t.Fatalf("Expected new target group but got none")
-			} else {
-				// Invalid type fsd should always break down.
-				break retry
-			}
-		case tgs := <-ch:
-			if !expect {
-				t.Fatalf("Unexpected target groups %s, we expected a failure here.", tgs)
+		case <-time.After(defaultWait):
+			t.Fatalf("Expected update but got none")
+			return
+		case <-time.After(defaultWait / 10):
+			if ref.Equal(t.lastReceive()) {
+				// No update received.
+				break
 			}
 
-			if len(tgs) != 2 {
-				continue retry // Potentially a partial write, just retry.
+			// We can receive partial updates so only check the result when the
+			// expected number of groups is reached.
+			tgs := t.targets()
+			if len(tgs) != len(expected) {
+				t.Logf("skipping update: expected %d targets, got %d", len(expected), len(tgs))
+				break
 			}
-			tg := tgs[0]
-
-			if _, ok := tg.Labels["foo"]; !ok {
-				t.Fatalf("Label not parsed")
-			}
-			if tg.String() != filepath.Join(testDir, "_test_"+prefix+ext+":0") {
-				t.Fatalf("Unexpected target group %s", tg)
-			}
-
-			tg = tgs[1]
-			if tg.String() != filepath.Join(testDir, "_test_"+prefix+ext+":1") {
-				t.Fatalf("Unexpected target groups %s", tg)
+			t.requireTargetGroups(expected, tgs)
+			if ref.After(time.Time{}) {
+				t.Logf("update received after %v", t.lastReceive().Sub(ref))
 			}
-			break retry
+			return
 		}
 	}
+}
 
-	// Based on unknown circumstances, sometimes fsnotify will trigger more events in
-	// some runs (which might be empty, chains of different operations etc.).
-	// We have to drain those (as the target manager would) to avoid deadlocking and must
-	// not try to make sense of it all...
-	drained := make(chan struct{})
-	go func() {
-		for {
-			select {
-			case tgs := <-ch:
-				// Below we will change the file to a bad syntax. Previously extracted target
-				// groups must not be deleted via sending an empty target group.
-				if len(tgs[0].Targets) == 0 {
-					t.Errorf("Unexpected empty target groups received: %s", tgs)
-				}
-			case <-time.After(500 * time.Millisecond):
-				close(drained)
-				return
+func (t *testRunner) requireTargetGroups(expected, got []*targetgroup.Group) {
+	t.Helper()
+	b1, err := json.Marshal(expected)
+	if err != nil {
+		panic(err)
+	}
+	b2, err := json.Marshal(got)
+	if err != nil {
+		panic(err)
+	}
+
+	testutil.Equals(t, string(b1), string(b2))
+}
+
+// validTg() maps to fixtures/valid.{json,yml}.
+func validTg(file string) []*targetgroup.Group {
```

Comment:
Whats the motivation behind checking the target groups via a function like this rather than expected return value vars in a test case struct?

Reply:
This isn't checking the target groups, just returning a `[]*targetgroup.Group`. The reason is that the `fileSDFilepathLabel` label depends on the location of the SD file so it is tedious to write in test case struct.